### PR TITLE
Restructured examples and documentation

### PR DESCRIPTION
**Pull request by SBottani**
Rewriting of the /examples folder
Debugging of density plot for plot.plot_neuron()
Updates in the doc

### CODE REVIEW

File: tests/test_examples.py [hunk 1]
```diff
@@ -24,57 +24,82 @@
...
 
 
-def test_1_first_steps(monkeypatch):
+def test_1_bipolar_cell(monkeypatch):
+    '''
+    Run first example.
+    '''
+    monkeypatch.setattr(plt, "show", mock_show)
+    ds.reset_kernel()
+    exec(open(models + "/bipolar_cell.py").read())
```
Comment (Silmathoron) [inline tests/test_examples.py:48]: this should probably be ``neuron_models``, see also below

File: docs/user/tutorial.rst [hunk 2]
```diff
@@ -227,6 +227,80 @@ Which leads to the following structure:
 .. image:: ./images/neurons_in_square.png
 
 
+Other Examples
+==============
+
+The ../examples folder contains a collections of different applications cases illustrating DeNSE functions.
```
[the remark this answers is not in the document]
Comment (SBottani) [reply docs/user/tutorial.rst:233]: Sorry I do not know how to do.

File: docs/user/recording.rst [hunk 3]
```diff
@@ -34,9 +34,22 @@ The file will be saved in the current working directory in the path:
...
 The functions to deal with data storage are written under the name DataIO.
 
-In order to save data you need to set the
+In order to save data you need to use the function of class "io".
+
+ex.:
+
++ dense.io.save_to_swc()
++ dense.io.save_json_info()
++ dense.io.save_to_neuroml()
```
[the remark this answers is not in the document]
Comment (SBottani) [reply docs/user/recording.rst:55]: ok

File: docs/user/recording.rst [hunk 4]
```diff
@@ -34,9 +34,22 @@ The file will be saved in the current working directory in the path:
 `<SimulationID>/record.dat`
 
 The `static_data` are the `morphology.swc` and `info.json`, both are produced at the end of the simulation, they can be placed where the user prefer, since we do expect these are the data usually required.
-For an insight on the SWC format you're pleased to read the apposite section.
 
+For an insight on the SWC format you're pleased to read the apposite section. The SWC format is one of the most widely used  neuron morphologyformats (in particular, a standardized version of this format is used by the neuromorpho_ http://www.neuromorpho.org/ archive  ).
```
[the remark this answers is not in the document]
Comment (SBottani) [reply docs/user/recording.rst:38]: ok

File: docs/user/recording.rst [hunk 5]
```diff
@@ -34,9 +34,22 @@ The file will be saved in the current working directory in the path:
...
+
+
+DENsE can also store neuron morphologies in the *neuroml* format, an alternative data format for defining and
```
[the remark this answers is not in the document]
Comment (SBottani) [reply docs/user/recording.rst:41]: ok

File: .gitignore [hunk 6]
```diff
@@ -1,5 +1,6 @@
 # Python
-
+ *.nml
+ *.swc
```
[the remark this answers is not in the document]
Comment (SBottani) [reply .gitignore:3]: ok

File: examples/multi_chambers/2chambers/2chambers.py [hunk 7]
```diff
@@ -119,85 +116,103 @@ def CleanFolder(tmp_dir, make=True):
...
     print("Creating neurons")
+<<<<<<< HEAD:examples/multi_chambers/2chambers/2chambers.py
+    gids = ds.create_neurons(n=num_neurons,
+                             culture=culture,
+                             params=neuron_params,
+                             neurite_params=dendrite_params,
+                             num_neurites=2)
+
+    print("neurons done")
+=======
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/multi_chambers/2chambers/2chambers.py:161]: done

File: examples/multi_chambers/2chambers/TODO [hunk 8]
```diff
@@ -0,0 +1,5 @@
+TODO
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/multi_chambers/2chambers/TODO:1]: done

File: src/pymodule/dense/io/data_swc.py [hunk 9]
```diff
@@ -69,7 +78,13 @@ def load_swc(swc_folder=None, swc_file=None, info=None):
...
         the format style of SWC files is the standard
 
-    @todo: redo this
+    @todo: redo this.
+            Currently if swc_file is given, the code assumes the file
+            contains only one neuron !
+            If both swc_folder and swc_file are given, it assumes the file
+            contains only one neuron.
+            If the swc file contains more thant one neurone, give only
+            the swc_folder path.
```
[the remark this answers is not in the document]
Comment (SBottani) [reply src/pymodule/dense/io/data_swc.py:87]: issue raised. 
But until resolved I suggest keeping these indications in the comments, as it is the current state  mode of operation.

File: tests/test_examples.py (hunk 1 above)
Comment (SBottani) [reply tests/test_examples.py:48]: indeed

File: examples/models/neurons/bipolar_cell.py [hunk 10]
```diff
@@ -91,7 +93,8 @@
 
 # first, elongation
 
-ds.simulate(7*day)
+ds.simulate(10000*minute)
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/models/neurons/bipolar_cell.py:96]: ok, strange I had this !

File: examples/models/neurons/bipolar_cell.py [hunk 11]
```diff
@@ -56,11 +57,12 @@
     # branching
     "use_van_pelt": True,
     "B": 0.2*cpm,
-    "T": 3.5*day,
-    "gc_split_angle_mean": 35.*deg,
+    "T": 5000.*minute,
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/models/neurons/bipolar_cell.py:60]: Indeed, it must have been from an old old file version

File: examples/models/competition/res_po_rt_neuron_exploration.py [hunk 12]
```diff
@@ -131,31 +154,31 @@ def run_dense(neuron_params):
...
+    step(1000./resolution * minute, 1, False, True)
+    step(1000./resolution * minute, 1, False, True)
+    step(1000./resolution * minute, 1, False, True)
+    step(1080 * minute, 1, False, True)
+    step(1080 * minute, 1, False, True)
+    step(1080 * minute, 1, False, True)
+    step(1080 * minute, 1, False, True)
+    step(1080 * minute, 1, False, True)
+    step(1080 * minute, 1, False, True)
+    step(4080 * minute, 1, False, True)
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/models/competition/res_po_rt_neuron_exploration.py:175]: I remove the whole fine.

It was so... gives an idea what kind of stuff was also in the folder...

File: examples/models/branching/branching_on_and_off_example.py [hunk 13]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 #
-# several_step_growth.py
+# Same code as /examples/models/neuron/several_step_growth.py
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/models/branching/branching_on_and_off_example.py:3]: ok

File: examples/growth_studies/turning_walls_exp/RenaudAngles.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    "filopodia_wall_affinity": 10.,
+    "persistence_length" : 200. * um,
+    "taper_rate": 3./250.,
+    'B': 10. * cpm,
+    'T': 10000. * minute,
+    'E': 0.7,
+}
+
+neurite_params = {"axon": axon_params}
+#neurite_params = {"axon": axon_params, "dendrite_1": dendrite_params}
```
[the remark this answers is not in the document]
Comment (SBottani) [reply examples/growth_studies/turning_walls_exp/RenaudAngles.py:94]: ok

File: examples/growth_studies/turning_walls_exp/RenaudAngles.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+min_x = 0  # "Arches20reduced2c2.svg" has x segment of length 164.91 cm
+max_x = 500  # and height of 84.94 cm
+min_y = -132
+max_y = 132  # set here same aspect ratio  X , Y as size of environment
+#  in the svg file
+culture_file = current_dir + "/angle40.svg"
+
+neuron_params = {"soma_radius": soma_radius,
+                 "random_rotation_angles": False,
+                 #"neurite_names": ["axon", "dendrite_1"],
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/growth_studies/turning_walls_exp/RenaudAngles.py]: the file has been suppressed, where is it now?

[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/growth_studies/turning_walls_exp/RenaudAngles.py]: funnel.py

**Comment by Silmathoron** [comment]
Il reste quelques erreurs sur les tests (probablement probabilistes ?) : [travis error](https://travis-ci.com/github/SENeC-Initiative/DeNSE/jobs/341277406#L1001) sur ``test_2_chandelier_cell`` (``chandelier-cell.py``).
À regarder en local si l'erreur existe aussi.

Il y a aussi une erreur d'import sur nngt dans ``several_step_growth_and_on_and_off_branching.py`` sur mac.

**Comment by SBottani** [comment]
test_2_chandelier_cell (chandelier-cell.py)
fonctionne localement chez moi. 

Enleve l'importation de nngt dans several_step_growth_and_on_and_off_branching.py 
car non utilisé !

**Comment by SBottani** [comment]
In the remaining failure of test_2_chandelier_cell (chandelier-cell.py)  does the travis report mean the issue is with plotting the dendogram ?
Works locally.
Wayout... have the dendogram optional?.... (I just want all this commited !)

**Comment by Silmathoron** [comment]
@SBottani I'm gonna check what may be wrong locally also, I'll keep you posted

**Comment by Silmathoron** [comment]
It's a real bug, I'll fix it then you can pull the fix and we can merge

**Comment by Silmathoron** [comment]
Please review #48, it fixes the bug so as soon as we merge it, you can pull and we should be good to go

**Comment by SBottani** [comment]
Hi Tanguy,
Sorry I was a bit lost and busy by other things. I approved the modified file in #48 but still need more testing. I do not know anymore where we are in fact.
S.

**Comment by Silmathoron** [comment]
@SBottani no worries, I merged #48 so if you include the changes on your branch by pulling master, tests should now hopefully pass

**Comment by Silmathoron** [comment]
Hey @SBottani it's almost good but there seem to be an issue with ``pip install nngt`` on Windows... I'll try to fix it today, then everything should work.
There should be nothing to do on your side.

File: examples/models/branching/several_step_growth_and_on_and_off_branching.py [hunk 16]
```diff
@@ -25,57 +28,57 @@
 import numpy as np
 import matplotlib.pyplot as plt
 
-import nngt
+#import nngt
```
Comment (Silmathoron) [inline examples/models/branching/several_step_growth_and_on_and_off_branching.py:31]: remove if unused

File: examples/models/branching/several_step_growth_and_on_and_off_branching.py [hunk 17]
```diff
@@ -141,26 +139,27 @@
...
+        "B": 5.* cpm,
+        "E": 0.1,
+        "S": 1.5, # large S leads to core dump
+        "T": 7*day,
     }
 
     neurite_params = {"axon": vp_axon, "dendrites": vp_dend}
 
     # update the parameters lists of the neurons 'gids'
+
```
Comment (Silmathoron) [inline examples/models/branching/several_step_growth_and_on_and_off_branching.py:162]: remove whiteline

File: examples/models/branching/several_step_growth_and_on_and_off_branching.py [hunk 18]
```diff
@@ -180,6 +179,7 @@
         "use_van_pelt": False,
         "use_uniform_branching": True,
         "uniform_branching_rate": 0.01 * cph,
+
```
Comment (Silmathoron) [inline examples/models/branching/several_step_growth_and_on_and_off_branching.py:182]: remove whiteline

File: examples/models/competition/critical.py [hunk 19]
```diff
@@ -198,50 +189,4 @@ def run_dense(neuron_params):
     # neuron1.plot_dendrogram()
     plt.show(block=True)
     # plt.savefig("dendrogram-E_{}-S_{}.pdf".format(E,S), format="pdf", ppi =300)
```
Comment (Silmathoron) [inline examples/models/competition/critical.py:191]: is this line necessary?
please also add a newline at the end of the file

File: examples/models/neurons/bipolar_cell.py [hunk 20]
```diff
@@ -20,33 +20,34 @@
 # along with DeNSE. If not, see <http://www.gnu.org/licenses/>.
 
 
-""" Generate the morphology of a bipolar cell (spindle neuron) """
+""" Generate the morphology of a bipolar cell (spindle neuron) 
+    Also illustrates generation of dendogram
+"""
```
Comment (Silmathoron) [inline examples/models/neurons/bipolar_cell.py:26]:
```suggestion
"""
Generate the morphology of a bipolar cell (spindle neuron) 
Also illustrates generation of dendogram
"""

```

File: examples/models/neurons/bipolar_cell.py [hunk 21]
```diff
@@ -20,33 +20,34 @@
...
 
 import numpy as np
 import os
 
 import matplotlib.pyplot as plt
 
+import neurom as nm
```
Comment (Silmathoron) [inline examples/models/neurons/bipolar_cell.py:32]: did we not say that we would encapsulate all parts requiring ``neurom`` into

```python
try:
    import neurom
    ...
except ImportError:
    pass
```

therefore doing the neurom import at the end of the file?

File: examples/models/neurons/bipolar_cell.py [hunk 22]
```diff
@@ -56,20 +57,21 @@
     # branching
     "use_van_pelt": True,
     "B": 0.2*cpm,
-    "T": 3.5*day,
-    "gc_split_angle_mean": 35.*deg,
+    "T": 5000.*minute,
```
Comment (Silmathoron) [inline examples/models/neurons/bipolar_cell.py:60]: Let's make it 3.5*day again

File: examples/models/neurons/chandelier-cell.py [hunk 23]
```diff
@@ -128,7 +129,13 @@
...
 }
 
-neurite_params = {"axon": axon_vp, "dendrites": dend_vp}
+
+axon_params.update(axon_vp)
+dend_params.update(dend_vp)
+dend_params.pop("initial_diameter")
+axon_params.pop("initial_diameter")
+dend_params.pop("taper_rate")
+axon_params.pop("taper_rate")
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/chandelier-cell.py:138]: up

File: examples/models/neurons/chandelier-cell.py [hunk 24]
```diff
@@ -141,29 +148,30 @@
...
+    "uniform_branching_rate": 0.05*cph,
     "lateral_branching_angle_mean": 45.*deg
 }
 
-neurite_params = {"axon": lb_a, "dendrites": lb}
+# neurite_params = {"axon": lb_a, "dendrites": lb}
+axon_params.update(lb_a)
+dend_params.update(lb)
 
 ds.set_object_properties(n, neurite_params=neurite_params)
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/chandelier-cell.py:168]: up

File: examples/models/neurons/chandelier-cell.py [hunk 25]
```diff
@@ -173,13 +181,21 @@
...
-
-ds.plot.plot_dendrogram(n.axon, show=False, vertical_diam_frac=0.45)
 ds.plot.plot_neurons(show=True)
 
 ds.io.save_to_swc("chandelier-cell.swc", gid=n, resolution=50)
 
-print("Asymmetry of axon:", ds.morphology.tree_asymmetry(n.axon))
+# deprecated syntax
+# tree = n.axon.get_tree()
+# tree.show_dendrogram()
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/chandelier-cell.py:194]: up

File: examples/models/neurons/granule_cell.py [hunk 26]
```diff
@@ -23,14 +23,18 @@
 """ Generate the morphology of a granule cell """
 
 import numpy as np
+import matplotlib.pyplot as plt
+
+import neurom as nm
```
Comment (Silmathoron) [inline examples/models/neurons/granule_cell.py:28]: encapsulation?

File: examples/models/neurons/multipolar-cell2.py [hunk 27]
```diff
@@ -52,12 +53,12 @@
...
     # Best model
-    "gc_split_angle_mean": 1.,
-    "B": 2.,
+    "gc_split_angle_mean": 30.*deg,
+    "B": 2.*cpm,
     "E": 0.,
     "S": 1.,
-    "T": 5000.,
-    "gc_split_angle_mean": 30.,
+    "T": 5000.*minute,
```
Comment (Silmathoron) [inline examples/models/neurons/multipolar-cell2.py:60]: 2.5*day

File: examples/models/neurons/multipolar-cell2.py [hunk 28]
```diff
@@ -66,54 +67,64 @@
...
 # Turn branching on
 
 #~ vp_branching = {'use_van_pelt': True}
-resource_branching = {'res_branching_threshold': 80., 'res_branching_proba': 0.0005}
-d_rsrc_branching = {'res_branching_threshold': 60., 'res_branching_proba': 0.0003}
+# resource_branching = {'res_branching_threshold': 80., 'res_branching_proba': 0.0005}
+# d_rsrc_branching = {'res_branching_threshold': 60., 'res_branching_proba': 0.0003}
 
 #~ ds.set_object_properties(n, params=vp_branching)
 #~ ds.set_object_properties(n, params=resource_branching)
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/multipolar-cell2.py:111]: up

File: examples/models/neurons/multipolar-cell2.py [hunk 29]
```diff
@@ -66,54 +67,64 @@
...
+dend_params['res_branching_threshold'] = 80.
+dend_params['res_branching_proba'] = 0.0003
+axon_params['res_branching_threshold'] = 80.
+axon_params['res_branching_proba'] = 0.0005
+
+dend_params.pop("initial_diameter")
+axon_params.pop("initial_diameter")
+dend_params.pop("taper_rate")
+axon_params.pop("taper_rate")
+
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/multipolar-cell2.py:122]: up

File: examples/models/neurons/multipolar-ressource-based.py [hunk 30]
```diff
@@ -112,107 +109,82 @@ def step(n, loop_n, save_path, plot=True):
...
-        "res_neurite_generated": 4500. * uM,
+        # "res_retraction_threshold": 0.4 * uM,
+        # "res_elongation_threshold": 0.15 * uM,
+        # "res_elongation_factor": 0.6 * um / minute,
+        # # 'use_van_pelt' : True,
+        # "res_neurite_generated": 4500. * uM,
+        # "res_neurite_delivery_tau": 50. * minute,
+        # "res_correlation": 0.15,
+        # "res_variance": 0.02 * uM / minute ** 0.5,
+        # "res_use_ratio": 0.3 * cpm,
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/multipolar-ressource-based.py:170]: up

File: examples/models/neurons/multipolar.py [hunk 31]
```diff
@@ -0,0 +1,191 @@
...
+        # "flpl_branching_rate" : 0.004,
+        # "res_retraction_threshold": 0.4 * uM,
+        # "res_elongation_threshold": 0.15 * uM,
+        # "res_elongation_factor": 0.6 * um / minute,
+        # # 'use_van_pelt' : True,
+        # "res_neurite_generated": 4500. * uM,
+        # "res_neurite_delivery_tau": 50. * minute,
+        # "res_correlation": 0.15,
+        # "res_variance": 0.02 * uM / minute ** 0.5,
+        # "res_use_ratio": 0.3 * cpm,
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/multipolar.py:171]: up

File: examples/models/neurons/pyramidal_cell.py [hunk 32]
```diff
@@ -24,16 +24,20 @@
 
 import numpy as np
 
+# import matplotlib as mpl
+# mpl.use("Qt5Agg")
+
```
[the remark this answers is not in the document]
Comment (Silmathoron) [reply examples/models/neurons/pyramidal_cell.py:29]: up

**Comment by SBottani** [comment]
Ok splendid !  (Splendid that *There should be nothing to do on your side.)*

See you soon
S.

Le lun. 29 juin 2020 à 10:02, Tanguy Fardet <notifications@github.com> a
écrit :

File: .travis.yml [hunk 33]
```diff
@@ -42,7 +39,7 @@ before_install:
     fi
   # update brew and pip on mac (note that mac is without nngt)
   - if [ "$TRAVIS_OS_NAME" = "osx" ]; then
-      brew install gcc@8 cmake python3 geos doxygen boost libomp;
+      brew install gcc cmake python3 geos doxygen boost@1.72 libomp;
```
Comment (Silmathoron) [inline .travis.yml:42]: Actually it seems the @ is necessary; can you update it again to ``@10`` this time

File: .travis.yml [hunk 34]
```diff
@@ -74,9 +71,9 @@ install:
   # on mac, set gcc as compiler
   - if [ "$TRAVIS_OS_NAME" = "osx" ]; then
       mkdir build && cd build;
-      CC=gcc-8 CXX=g++-8 cmake .. -Dwith-python=3;
+      CC=gcc CXX=g++ cmake .. -Dwith-python=3;
```
Comment (Silmathoron) [inline .travis.yml:74]: move to ``gcc-10`` and `g++-10`, same below

File: examples/models/neurons/granule_cell.py [hunk 35]
```diff
@@ -113,12 +110,19 @@
...
+
+    nrn = nm.load_neuron("granule-cell.swc")
+
+    fig, _ = viewer.draw(nrn)
 
-for ax in fig.axes:
-    ax.set_title("")
+    for ax in fig.axes:
+        ax.set_title("")
+except import error:
```
Comment (Silmathoron) [inline examples/models/neurons/granule_cell.py:124]: it's ``ImportError``, correct everywhere

File: .travis.yml [hunk 36]
```diff
@@ -87,8 +83,9 @@ script:
...
-      python3 -m pytest tests --ignore=tests/todo;
-    else
-      python -m pytest tests --ignore=tests/todo;
-    fi;
+  # - if [ "$TRAVIS_OS_NAME" = "osx" ]; then
+  #     python3 -m pytest tests --ignore=tests/todo;
+  #   else
+  #     python -m pytest tests --ignore=tests/todo;
+  #   fi;
+    python -m pytest tests --ignore=tests/todo;
```
Comment (Silmathoron) [inline .travis.yml:91]:
```suggestion
  - python -m pytest tests --ignore=tests/todo;

```

**Review by Silmathoron — APPROVED** [review]

**Pull request merged**